src/solvers/scip.rs

```diff
@@ -0,0 +1,190 @@
+        .hide_output()
+        .include_default_plugins()
+        .create_prob("problem")
+        .set_obj_sense(match to_solve.direction {
+            ObjectiveDirection::Maximisation => ObjSense::Maximize,
+            ObjectiveDirection::Minimisation => ObjSense::Minimize,
+        });
+    let mut var_map = HashMap::new();
+
+    for (
+        var,
+        &VariableDefinition {
+            min,
+            max,
+            is_integer,
+            ref name,
+        },
+    ) in to_solve.variables.iter_variables_with_def()
+    {
+        let coeff = *to_solve
+            .objective
+            .linear
+            .coefficients
+            .get(&var)
+            .unwrap_or(&0.);
+        let var_type = match is_integer {
+            true => VarType::Integer,
+            false => VarType::Continuous,
+        };
+        let id = model.add_var(min, max, coeff, name.as_str().clone(), var_type);
+        var_map.insert(var, id);
+    }
+
+    SCIPProblem {
+        model: model,
+        id_for_var: var_map,
+    }
+}
+
+/// A SCIP Model
+pub struct SCIPProblem {
+    // the underlying SCIP model representing the problem
+    model: Model<ProblemCreated>,
+    // map from good_lp variables to SCIP variable ids
+    id_for_var: HashMap<Variable, usize>,
+}
+
+impl SCIPProblem {
+    /// Get access to the raw russcip model
+    pub fn as_inner(&self) -> &Model<ProblemCreated> {
+        &self.model
+    }
+
+    /// Get mutable access to the raw russcip model
+    pub fn as_inner_mut(&mut self) -> &mut Model<ProblemCreated> {
+        &mut self.model
+    }
+}
+
+impl SolverModel for SCIPProblem {
+    type Solution = SCIPSolved;
+    type Error = ResolutionError;
+
+    fn solve(self) -> Result<Self::Solution, Self::Error> {
+        let solved_model = self.model.solve();
+        let status = solved_model.get_status();
+        match status {
+            russcip::status::Status::Optimal => Ok(SCIPSolved {
+                solved_problem: solved_model,
+                id_for_var: self.id_for_var,
+            }),
+            russcip::status::Status::Infeasible => {
+                return Err(ResolutionError::Infeasible);
+            }
+            russcip::status::Status::Unbounded => {
+                return Err(ResolutionError::Unbounded);
+            }
+            other_status => {
+                return Err(ResolutionError::Str(format!(
+                    "Unexpected status {:?}",
+                    other_status
+                )));
+            }
+        }
+    }
+
+    fn add_constraint(&mut self, c: Constraint) -> ConstraintReference {
+        let constant = -c.expression.constant;
+        let lhs = match c.is_equality {
+            true => constant,
+            false => -f64::INFINITY,
+        };
+
+        let n_vars_in_cons = c.expression.linear.coefficients.len();
+        let mut vars_in_cons = Vec::with_capacity(n_vars_in_cons);
+        let mut coeffs = Vec::with_capacity(n_vars_in_cons);
```

Comment:
This is ok; but in the future, we might envision being able to create scip constraint objects directly through russcip, and avoid the creation of these potentially large intermediary arrays.